keras_cv/losses/dice.py

```diff
@@ -0,0 +1,261 @@
+
+    # Calculate the weighted dice score and normalizer.
+    weighted_numerator = tf.reduce_sum(weight * numerator)
+    weighted_denominator = tf.reduce_sum(weight * denominator)
+    general_dice_score = weighted_numerator / weighted_denominator
+
+    return general_dice_score
+
+
+def _adaptive_dice_score(numerator, denominator):
+    # Calculate the dice scores
+    dice_score = numerator / denominator
+    # Calculate weights based on Dice scores.
+    weights = tf.exp(-1.0 * dice_score)
+    # Multiply weights by corresponding scores and get sum.
+    weighted_dice = tf.reduce_sum(weights * dice_score)
+    # Calculate normalization factor.
+    normalizer = tf.cast(tf.size(input=dice_score), dtype=tf.float32) * tf.exp(-1.0)
+    # normalize the dice score
+    norm_dice_score = weighted_dice / normalizer
+
+    return norm_dice_score
+
+
+def _check_input_params(beta, loss_type=None, class_ids=None):
+    if beta <= 0.0:
+        raise ValueError(f"`beta` value should be greater than zero. Got {beta}")
+
+    if loss_type is not None:
+        if loss_type not in ["generalized", "adaptive"]:
+            raise ValueError(
+                "The `loss_type` is not valid. "
+                "If `loss_type` is not `None`, It should be either "
+                f"`generalized` or `adaptive`. Got {loss_type}"
+            )
+
+    if class_ids is not None:
+        if isinstance(class_ids, float) or any(isinstance(x, float) for x in class_ids):
+            raise ValueError(
+                f"The indices should be int or a list of integer. Got {class_ids}"
+            )
+
+
+@tf.keras.utils.register_keras_serializable(package="keras_cv")
+class SparseDice(tf.keras.losses.Loss):
+    """Compute the dice loss between the sparse labels and predictions.
+
+    Use this sparse dice loss function when there are 2D and 3D semantic
+    segmentaiton task. We expect labels to be provided in a `sparse`
+    representation (integer-encoded).
+
+    In the snippet below, there are `num_classes` times channels per
+    sample. The shape of `y_true` and `y_pred` are
+    `[batch_size, height, width, class]` and
+    `[batch_size, height, width, num_classes]`.
+
+    Standalone usage:
+
+    >>> y_true = tf.random.uniform([5, 10, 10, 1], 0, maxval=4, dtype=tf.int32)
+    >>> y_pred = tf.random.uniform([5, 10, 10, 4], 0, maxval=4)
+    >>> dice = SparseDice()
+    >>> dice(y_true, y_pred).numpy()
+    0.5692612
+
+    >>> # Calling with 'sample_weight'.
+    >>> dice(y_true, y_pred, sample_weight=tf.constant([[0.5, 0.5]])).numpy()
+    0.2846306
+
+    Usage with the `compile()` API:
+    ```python
+    model.compile(optimizer='adam', loss=keras_cv.losses.SparseDice())
+    ```
+    """
+
+    def __init__(
+        self,
+        beta=1,
+        from_logits=False,
+        class_ids=None,
+        axis=[1, 2],
+        loss_type=None,
+        label_smoothing=0.0,
+        epsilon=1e-07,
+        per_image=False,
```

Comment:
> Note: I haven't been able to test the per_image argument. @innat mentioned that it provides a performance boost in some cases, but I'm not sure which tasks these are. Could you share some that we can benchmark the argument?

https://github.com/keras-team/keras-cv/issues/296#issuecomment-1326053384